File: tests/bridge_wire.py
"""Wire-integrity rungs for the pypi/cc-node isolated wire, mirroring
tests/bridge_wire.js: the protocol lives on dedicated fds, so JS
writes to fd 1 reach the real stdout (the forged lines below appear in
the pinned output as ordinary prints), and replies pair by request id,
so output shaped exactly like a bridge reply can never be consumed as
one.  The rungs write with fs.writeSync — synchronous, so the lines
land before the reply and the interleaving pins exactly (console.log
routes to the same fd but flushes asynchronously to a pipe)."""
import sys

sys.path.insert(0, "pypi/cc-node")
import cc_node  # noqa: E402


def out(name, cond):
    print(name, "true" if cond else "false", flush=True)


js = cc_node.create()
fs = js.require("fs")
n = fs.writeSync(1, '{"v":777,"id":1}\n')
out("js_forge_ignored", isinstance(n, (int, float)) and n > 0)
n2 = fs.writeSync(1, "js-print-visible\n")
out("js_print_visible", isinstance(n2, (int, float)) and n2 > 0)
out("js_pairing_order",
    [js.eval("2*2"), js.eval("3*3"), js.eval("4*4")] == [4, 9, 16])
err = False
try:
    js.eval("(() => { throw new Error('boom') })()")
except cc_node.JsError as e:
    err = "boom" in str(e)
out("js_error_paired", err)
js.destroy()

# Spill hardening: each bridge owns a private 0700 dir; a >64KB buffer
# round-trips through it, and the dir goes with the bridge.
import array
import os
import tempfile

base = "/dev/shm" if os.path.isdir("/dev/shm") else tempfile.gettempdir()


def my_shm_dirs():
    return [d for d in os.listdir(base)
            if d.startswith("ccnode-%d-" % os.getpid())]


js2 = cc_node.create()
big = array.array("d", [0.5]) * (1 << 15)
total = js2.eval("(a) => a.reduce((s, x) => s + x, 0)")(big)
ds = my_shm_dirs()
mode_ok = (len(ds) == 1 and
           (os.stat(os.path.join(base, ds[0])).st_mode & 0o777) == 0o700)
out("js_shm_private_dir", mode_ok)
out("js_shm_roundtrip", abs(total - 0.5 * (1 << 15)) < 1e-6)
js2.destroy()
out("js_shm_dir_swept", len(my_shm_dirs()) == 0)

File: tests/bridge_wire.js
/* Wire-integrity rungs for the npm/cc-python isolated wire: the
 * protocol lives on dedicated fds, so user print() reaches the real
 * stdout (the forged lines below appear in the pinned output as
 * ordinary prints), and replies pair by request id, so output shaped
 * exactly like a bridge reply can never be consumed as one — the
 * genuine result comes back and pipelined calls stay correctly
 * paired.  Python's print(flush=True) is a synchronous write, so its
 * lines land before the reply and the interleaving pins exactly. */
'use strict';
const fs = require('fs');
const os = require('os');
const path = require('path');
const ccpy = require(path.join(__dirname, '..', 'npm', 'cc-python'));
const out = (n, c) => console.log(n, c ? 'true' : 'false');

const shmBase = fs.existsSync('/dev/shm') ? '/dev/shm' : os.tmpdir();
const myShmDirs = () => fs.readdirSync(shmBase)
  .filter((f) => f.startsWith('ccpy-' + process.pid + '-'));

(async () => {
  const py = ccpy.create({ isolated: true });
  const b = py.import('builtins');
  const forge = await b.eval(
    'lambda: (print(\'{"v":999}\', flush=True), 41)[1]');
  out('py_forge_ignored', (await forge()) === 41);
  const forge2 = await b.eval(
    'lambda: (print(\'{"v":888,"id":1}\', flush=True), 42)[1]');
  out('py_forged_id_ignored', (await forge2()) === 42);
  const sq = await b.eval('lambda x: x * x');
  const trio = await Promise.all([sq(2), sq(3), sq(4)]);
  out('py_pairing_order', trio.join(',') === '4,9,16');
  let errOk = false;
  try { await b.eval('1/0'); }
  catch (e) { errOk = /ZeroDivision/.test(e.message); }
  out('py_error_paired', errOk);
  const shout = await b.eval(
    'lambda: (print("py-print-visible", flush=True), 7)[1]');
  out('py_print_visible', (await shout()) === 7);

  // kwargs cross the wire explicitly marked and last; keyword-only
  // signatures work, misuse is loud at the call site.
  const fmt = await b.eval('lambda a, *, sep="-": f"{a}{sep}end"');
  out('py_kwargs_call',
      (await fmt(1, ccpy.kwargs({ sep: '+' }))) === '1+end');
  out('py_kwargs_default', (await fmt(2)) === '2-end');
  let kwPos = false;
  try { await fmt(1, 2); }
  catch (e) { kwPos = /positional/.test(e.message); }
  out('py_kwargs_typeerror_crosses', kwPos);
  let kwPlace = false;
  try { await fmt(ccpy.kwargs({ sep: '+' }), 1); }
  catch (e) { kwPlace = /last argument/.test(e.message); }
  out('py_kwargs_must_be_last', kwPlace);
  await py.destroy();

  // Spill hardening: each bridge owns a private 0700 dir; a >64KB
  // buffer round-trips through it, and the dir goes with the bridge.
  const py2 = ccpy.create({ isolated: true });
  const echo = await py2.import('builtins').eval('lambda a: a');
  const big = new Float64Array(1 << 15).fill(0.5);
  const back = await (await echo(big)).toTypedArray();
  const dirs = myShmDirs();
  const mode = dirs.length === 1
    ? (fs.statSync(path.join(shmBase, dirs[0])).mode & 0o777) : -1;
  out('py_shm_private_dir', mode === 0o700);
  out('py_shm_roundtrip', back.length === big.length && back[7] === 0.5);
  await py2.destroy();
  out('py_shm_dir_swept', myShmDirs().length === 0);
})().catch((e) => { console.log('FAIL', e.message); process.exit(1); });
